target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c

```diff
@@ -2173,33 +2175,46 @@ static int rtl838x_mdio_init(struct rtl838x_eth_priv *priv)
 	for_each_node_by_name(dn, "ethernet-phy") {
 		u32 smi_addr[2];
 
-		if (of_property_read_u32(dn, "reg", &pn))
+		if (of_property_read_u32(dn, "reg", &pn)) {
+			pr_err("%s: missing reg property on port %d\n", __func__, pn);
 			continue;
+		}
 
-		if (of_property_read_u32_array(dn, "rtl9300,smi-address", &smi_addr[0], 2)) {
-			smi_addr[0] = 0;
-			smi_addr[1] = pn;
+		if (pn >= MAX_PORTS) {
+			pr_err("%s: illegal port number %d\n", __func__, pn);
+			continue;
 		}
 
+		priv->phy_is_internal[pn] = of_property_read_bool(dn, "phy-is-integrated");
+
 		if (of_property_read_u32(dn, "sds", &priv->sds_id[pn]))
 			priv->sds_id[pn] = -1;
 		else {
 			pr_info("set sds port %d to %d\n", pn, priv->sds_id[pn]);
 		}
 
-		if (pn < MAX_PORTS) {
+		if (of_property_read_u32_array(dn, "rtl9300,smi-address", &smi_addr[0], 2)) {
+			/* Integrated PHYs associated to a SerDes do not have an smi_bus */
+			if (priv->phy_is_internal[pn] && priv->sds_id[pn] >= 0) {
+				priv->smi_bus[pn] = -1;
```

Comment:
Setting a u32 to -1, are we? ;)

In any case, I don't think you should even try reading this property if you know you're not going to use it for the current port.

Reply:
hah, fair point; but, I think this is still a legal (C) construct. You are basically setting it to u32_max; but I think I saw that macro somewhere, so we can use that macro instead.

And that's exactly what this code does, if we don't have the property, we need to set the marker, as the `smi_bus` property is being used all over the code.

I don't intend to 'change the world' with this change, and it's birgers change of course, as he wrote all of this anyway.

Reply:
The problem with Birger's patches was, and still is, code quality. I'm fine with the rationale of this patch set, but the implementation needs to improve before I'm merging this.

Reply:
That's a bit over the top yeah? This all will be thrown away, this is not usable in its current form, it needs to be completely rewritten. We lack full understanding of a lot of things to be able to fully re-write it. The code quality of this patch is more then adequate, considering what's underneath. Gold-plating a turd won't help anyone.

Obvious mistakes, absolutely. But we are swimming in a pile of diarrhea, and you want to spend enormous amounts of time, on something that will never end up upstream (because its just wrong in many ways, wrong location, wrong subsystem, wrong 'type' of driver). I don't know how much spare time you have for all of this, but the rest of us _are_ limited herein.

If this where something that would possibly go upstream, sure, we can do better.

But the purpose of this, is to gain users, fix bugs, get more functionality and learn how the hardware works. This allows us to write actual upstream able code.

What we currently have upstream, has been problematic already (bugs in gpio for example) due to us not understanding the hardware. The drivers upstream even need to be improved a little. The SPI driver for example, while basic and functional, lacks a lot still. And that is a 200 line driver.

With this attitude, we won't ever finish support for these switches honestly. And we seriously lack manpower (and understanding of the hardware) to even consider going the 'pure mainline route'. Ideal, yes, realistic ... you guys have been at this for 3 years. Is there anybody contributing or USING mainline?

Babysteps are needed really. Otherwise, we just end up alienating developers (which has been happening all over, and yes, that code quality was poor, yet we merged the clock, timer etc drivers, which also are far from ideal).

In short; we have what we have. It's ugly, its bad. But don't let perfect be the enemy of better. Small betters is what we need for now. We'd love to make everything great again. Baby. Steps.